README


This archive contains the distribution DBIx-Async,
version 0.002:

  database support for IO::Async via DBI

This software is copyright (c) 2012 by Tom Molesworth.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.


This README file was generated by Dist::Zilla::Plugin::Readme v5.009.



Changes
Revision history for DBIx-Async

0.002     2014-01-13 09:34:38+00:00 Europe/London
	No new features.

	Bug fixes:
	* Minimum Future version is 0.21, for ->is_done
	* Don't assume DBD::SQLite is available (was being loaded in DBIx::Async::Worker::SQLite compilation tests)
	* Don't special-case sqlite setup (WAL journal etc.)

0.001     2014-01-12 19:09:03+00:00 Europe/London
	Initial CPAN release


lib/DBIx/Async.pm
package DBIx::Async;
# ABSTRACT: database support for IO::Async via DBI
use strict;
use warnings;
use parent qw(IO::Async::Notifier);

our $VERSION = '0.002';

=head1 NAME

DBIx::Async - use L<DBI> with L<IO::Async>

=head1 VERSION

version 0.002

=head1 SYNOPSIS

 #!/usr/bin/env perl 
 use strict;
 use warnings;
 use feature qw(say);
 use IO::Async::Loop;
 use DBIx::Async;
 my $loop = IO::Async::Loop->new;
 say 'Connecting to db';
 $loop->add(my $dbh = DBIx::Async->connect(
   'dbi:SQLite:dbname=test.sqlite3',
   '',
   '', {
     AutoCommit => 1,
     RaiseError => 1,
   }
 ));
 $dbh->do(q{CREATE TABLE tmp(id integer primary key autoincrement, content text)})
 # ... put some values in it
 ->then(sub { $dbh->do(q{INSERT INTO tmp(content) VALUES ('some text'), ('other text') , ('more data')}) })
 # ... and then read them back
 ->then(sub {
   # obviously you'd never really use * in a query like this...
   my $sth = $dbh->prepare(q{select * from tmp});
   $sth->execute;
   # the while($row = fetchrow_hashref) construct isn't a good fit
   # but we attempt to provide something reasonably close using the
   # ->iterate helper
   $sth->iterate(
     fetchrow_hashref => sub {
       my $row = shift;
       say "Row: " . join(',', %$row);
     }
   );
 })->on_done(sub {
   say "Query complete";
   $loop->stop;
 })->on_fail(sub { warn "Failure: @_\n" });
 $loop->run;

=head1 DESCRIPTION

Wrapper for L<DBI>, for running queries much slower than usual but without blocking.

C<NOTE>: This is an early release, please get in contact via email (see L</AUTHOR>
section) or RT before relying on it for anything.

=head2 PERFORMANCE

Greatly lacking. See C<examples/benchmark.pl>, in one sample run the results looked
like this:

               Rate DBIx::Async DBD::SQLite
 DBIx::Async 1.57/s          --        -89%
 DBD::SQLite 13.8/s        776%          --

If you're doing anything more than occasional light queries, you'd probably be better
off with blocking DBI-based code running in a fork.

=head1 METHODS

Where possible, L<DBI> method signatures are used.

=cut

use IO::Async::Channel;
use IO::Async::Routine;
use Future;
use Module::Load qw();
use Try::Tiny;

use DBIx::Async::Handle;

use constant DEBUG => 0;

=head2 connect

Constuctor. Sets up our instance with parameters that will be used when we attempt to
connect to the given DSN.

Takes the following options:

=over 4

=item * $dsn - the data source name, should be something like 'dbi:SQLite:dbname=:memory:'

=item * $user - username to connect as

=item * $pass - password to connect with

=item * $opt - any options

=back

Options consist of:

=over 4

=item * RaiseError - set this to 1

=item * AutoCommit - whether to run in AutoCommit mode by default, probably works better
if this is set to 1 as well

=back

C<NOTE>: Despite the name, this method does not initiate a connection. This may change in
a future version, but if this behaviour does change this method will still return C<$self>.

Returns $self.

=cut

sub connect {
	my $class = shift;
	my ($dsn, $user, $pass, $opt) = @_;
	my $self = bless {
		options => {
			RaiseError => 1,
			PrintError => 0,
			AutoCommit => 1,
			%{ $opt || {} },
		},
		pass    => $pass,
		user    => $user,
		dsn     => $dsn,
	}, $class;
	$self
}

=head2 dsn

Returns the DSN used in the L</connect> request.

=cut

sub dsn { shift->{dsn} }

=head2 user

Returns the username used in the L</connect> request.

=cut

sub user { shift->{user} }

=head2 pass

Returns the password used in the L</connect> request.

=cut

sub pass { shift->{pass} }

=head2 options

Returns any options that were set in the L</connect> request.

=cut

sub options { shift->{options} }

=head2 do

Runs a query with optional bind parameters. Takes the following parameters:

=over 4

=item * $sql - the query to run

=item * $options - any options to apply (can be undef)

=item * @params - the parameters to bind (can be empty)

=back

Returns a L<Future> which will resolve when this query completes.

=cut

sub do : method {
	my ($self, $sql, $options, @params) = @_;
	$self->queue({
		op => 'do',
		sql => $sql,
		options => $options,
		params => \@params,
	});
}

=head2 begin_work

Starts a transaction.

Returns a L<Future> which will resolve when this transaction has started.

=cut

sub begin_work {
	my $self = shift;
	$self->queue({ op => 'begin_work' });
}

=head2 commit

Commit the current transaction.

Returns a L<Future> which will resolve when this transaction has been committed.

=cut

sub commit {
	my $self = shift;
	$self->queue({ op => 'commit' });
}

=head2 savepoint

Marks a savepoint. Takes a single parameter: the name to use for the savepoint.

 $dbh->savepoint('here');

Returns a L<Future> which will resolve once the savepoint has been created.

=cut

sub savepoint {
	my $self = shift;
	my $savepoint = shift;
	$self->queue({ op => 'savepoint', savepoint => $savepoint });
}


=head2 release

Releases a savepoint. Takes a single parameter: the name to use for the savepoint.

 $dbh->release('here');

This is similar to L</commit> for the work which has been completed since
the savepoint, although the database state is not updated until the transaction
itself is committed.

Returns a L<Future> which will resolve once the savepoint has been released.

=cut

sub release {
	my $self = shift;
	my $savepoint = shift;
	$self->queue({ op => 'release', savepoint => $savepoint });
}

=head2 rollback

Rolls back this transaction. Takes an optional savepoint which
can be used to roll back to the savepoint rather than cancelling
the entire transaction.

Returns a L<Future> which will resolve once the transaction has been
rolled back.

=cut

sub rollback {
	my $self = shift;
	my $savepoint = shift;
	$self->queue({ op => 'rollback', savepoint => $savepoint });
}

=head2 prepare

Attempt to prepare a query.

Returns the statement handle as a L<DBIx::Async::Handle> instance.

=cut

sub prepare {
	my $self = shift;
	my $sql = shift;
	DBIx::Async::Handle->new(
		dbh => $self,
		prepare => $self->queue({ op => 'prepare', sql => $sql }),
	);
}

=head1 INTERNAL METHODS

These are unlikely to be of much use in application code.

=cut

=head2 queue

Queue a request. Used internally.

Returns a L<Future>.

=cut

sub queue {
	my $self = shift;
	my ($req, $code) = @_;
	my $f = $self->loop->new_future;
	if(DEBUG) {
		require Data::Dumper;
		warn "Sending req " . Data::Dumper::Dumper($req);
	}
	$self->sth_ch->send($req);
	$self->ret_ch->recv(
		on_recv => sub {
			my ( $ch, $rslt ) = @_;
			if($rslt->{status} eq 'ok') {
				$f->done($rslt);
			} else {
				$f->fail($rslt->{message});
			}
		}
	);
	$f
}

=head2 worker_class_from_dsn

Returns $self.

=cut

sub worker_class_from_dsn {
	my $self = shift;
	my $dsn = shift;
	my ($dbd) = $dsn =~ /^dbi:([^:]+)(?::|$)/;
	die "Invalid DBD class: $dbd" unless $dbd =~ /^[a-zA-Z0-9]+$/;
	my $loaded;
	my $class;
	for my $subclass ($dbd, 'Default') {
		last if $loaded;
		$class = 'DBIx::Async::Worker::' . $subclass;
		try {
			Module::Load::load($class);
			$loaded = 1
		} catch {
			warn "class load: $_\n" if DEBUG;
		};
	}
	die "Could not find suitable class for $dbd" unless $loaded;
	$class;
}

=head2 sth_ch

The channel used for prepared statements.

=cut

sub sth_ch { shift->{sth_ch} }

=head2 ret_ch

The channel which returns values.

=cut

sub ret_ch { shift->{ret_ch} }

=head2 _add_to_loop

Sets things up when we are added to a loop.

=cut

sub _add_to_loop {
	my ($self, $loop) = @_;

	my $worker = $self->worker_class_from_dsn($self->dsn)->new(
		parent => $self,
		sth_ch => ($self->{sth_ch} = IO::Async::Channel->new),
		ret_ch => ($self->{ret_ch} = IO::Async::Channel->new),
	);
	my $routine = IO::Async::Routine->new(
		channels_in  => [ $self->{sth_ch} ],
		channels_out => [ $self->{ret_ch} ],
		code => sub { $worker->run },
		on_finish => sub {
			print "The routine aborted early - $_[-1]\n";
			$self->loop->stop;
		},
	);
	$loop->add($routine);
}

=head2 _remove_from_loop

Doesn't do anything.

=cut

sub _remove_from_loop {
	my $self = shift;
	my ($loop) = @_;
	warn "Removed from loop\n" if DEBUG;
}

1;

__END__

=head1 TODO

=over 4

=item * Much of the L<DBI> API is not yet implemented. Fix this.

=item * Provide a nicer wrapper around transactions.

=item * Consider supporting L<Net::Async::PostgreSQL> and L<Net::Async::MySQL>
natively, might lead to some performance improvements.

=back

=head1 SEE ALSO

=over 4

=item * L<DBI> - the database framework that does all the real work

=item * L<Net::Async::PostgreSQL> - nonblocking interaction with PostgreSQL, not DBI compatible

=back

=head1 AUTHOR

Tom Molesworth <cpan@entitymodel.com>

=head1 LICENSE

Copyright Tom Molesworth 2012-2014. Licensed under the same terms as Perl itself.


lib/DBIx/Async/Handle.pm
package DBIx::Async::Handle;
{
  $DBIx::Async::Handle::VERSION = '0.002';
}
use strict;
use warnings;

=head1 NAME

DBIx::Async::Handle - statement handle for L<DBIx::Async>

=head1 VERSION

version 0.002

=head1 DESCRIPTION

Some L<DBIx::Async> methods (L<DBIx::Async/prepare> for example)
return statement handles. Those statement handles are supposed to
behave something like L<DBI>'s statement handles. Where they don't,
this is either a limitation of the async interface or a bug. Please
report if the latter.

=head1 METHODS

=cut

=head2 new

Returns $self.

=cut

sub new { my $class = shift; bless { @_ }, $class }

=head2 dbh

Returns the database handle which created this statement handle.

=cut

sub dbh { shift->{dbh} }

=head2 execute

Executes this statement handle, takes an optional list of bind parameters.

Returns a L<Future> which will resolve when the statement completes.

=cut

sub execute {
	my $self = shift;
	my @param = @_;
	$self->{execute} = $self->{prepare}->then(sub {
		my $id = shift->{id};
		$self->dbh->queue({ op => 'execute', id => $id, param => \@param });
	});
}

=head2 finish

Marks this statement handle as finished.

Returns a L<Future> which will resolve when the statement is finished.

=cut

sub finish {
	my $self = shift;
	my @param = @_;
	die "execute has not yet completed?" unless $self->{execute}->is_ready;
	$self->{execute}->then(sub {
		my $id = shift->{id};
		$self->dbh->queue({ op => 'finish', id => $id });
	});
}

=head2 fetchrow_hashref

Fetch a single row, returning the results as a hashref.

Since the data won't necessarily be ready immediately, this returns
a L<Future> which will resolve with the requested hashref.

=cut

sub fetchrow_hashref {
	my $self = shift;
	die "fetch on a handle which has not been executed" unless $self->{execute};
	$self->{execute}->then(sub {
		my $id = shift->{id};
		$self->dbh->queue({
			op => 'fetchrow_hashref',
			id => $id
		});
	})->transform(done => sub { shift->{data} // () })
}

=head2 iterate

A helper method for iterating over results.

Takes two parameters:

=over 4

=item * $method - the method to call, for example L</fetchrow_hashref>

=item * $code - the code to run for each result

=back

 $sth->iterate(
  fetchrow_hashref => sub {
   
  }
 )

Returns $self.

=cut

sub iterate {
	my $self = shift;
	my $method = shift;
	my $code = shift;
	my $f = $self->dbh->loop->new_future;
	my $step;
	$step = sub {
		return $f->done unless @_;
		$self->$method->on_done($step);
		$code->(@_);
	};
	$self->$method->on_done($step);
	$f;
}

1;

__END__

=head1 TODO

=over 4

=item * There are many other ->fetch* variants. Add them.

=back

=head1 AUTHOR

Tom Molesworth <cpan@entitymodel.com>

=head1 LICENSE

Copyright Tom Molesworth 2012-2014. Licensed under the same terms as Perl itself.


lib/DBIx/Async/Worker.pm
package DBIx::Async::Worker;
{
  $DBIx::Async::Worker::VERSION = '0.002';
}
use strict;
use warnings;

=head1 NAME

DBIx::Async::Worker - background process for L<DBIx::Async>

=head1 VERSION

version 0.002

=head1 DESCRIPTION

No user-serviceable parts inside. You may want to subclass this
for specific DBD driver support though.

=cut

use DBI;
use Try::Tiny;

my %VALID_METHODS;
BEGIN {
	@VALID_METHODS{qw(
		begin_work commit rollback savepoint release
		do
		prepare execute finish
		fetchrow_hashref
	)} = ();
}

my %sth;

=head1 METHODS

=cut

=head2 new

Returns $self.

=cut

sub new { my $class = shift; bless { @_ }, $class }

=head2 ret_ch

Returns $self.

=cut

sub ret_ch { shift->{ret_ch} }

=head2 sth_ch

Returns $self.

=cut

sub sth_ch { shift->{sth_ch} }

=head2 parent

Returns $self.

=cut

sub parent { shift->{parent} }

sub connect {
	my $self = shift;
	$self->{dbh} = DBI->connect(
		$self->parent->dsn,
		$self->parent->user,
		$self->parent->pass,
		$self->parent->options
	);
	$self;
}

sub do : method {
	my $self = shift;
	my $op = shift;
	$self->dbh->do(
		$op->{sql},
		$op->{options},
		@{ $op->{params} || [] }
	);
	return { status => 'ok' };
}

sub begin_work {
	my $self = shift;
	my $op = shift;
	$self->dbh->begin_work;
	return { status => 'ok' };
}

sub commit {
	my $self = shift;
	my $op = shift;
	$self->dbh->commit;
	return { status => 'ok' };
}

sub savepoint {
	my $self = shift;
	my $op = shift;
	$self->dbh->do(q{savepoint} . (defined $op->{savepoint} ? ' ' . $self->dbh->quote_identifier($op->{savepoint}) : ''));
	return { status => 'ok' };
}

sub release {
	my $self = shift;
	my $op = shift;
	$self->dbh->do(q{release savepoint } . $self->dbh->quote_identifier($op->{savepoint}));
	return { status => 'ok' };
}

sub rollback {
	my $self = shift;
	my $op = shift;
	$self->dbh->rollback;
	return { status => 'ok' };
}

sub prepare {
	my $self = shift;
	my $op = shift;
	my $sth = $self->dbh->prepare($op->{sql});
	$sth{$sth} = $sth;
	return { status => 'ok', id => "$sth" };
}

sub finish {
	my $self = shift;
	my $op = shift;
	my $sth = delete $sth{$op->{id}} or return {
		status => 'fail',
		message => 'invalid ID'
	};
	$sth->finish;
	return { status => 'ok', id => "$sth" };
}

sub execute {
	my $self = shift;
	my $op = shift;
	my $sth = $sth{$op->{id}} or return {
		status => 'fail',
		message => 'invalid ID'
	};
	$sth->execute(@{ $op->{param} });
	return { status => 'ok', id => "$sth" };
}

sub fetchrow_hashref {
	my $self = shift;
	my $op = shift;
	my $sth = $sth{$op->{id}} or return {
		status => 'fail',
		message => 'invalid ID'
	};
	my $data = $sth->fetchrow_hashref;
	return { status => 'ok', id => "$sth", data => $data };
}

sub dbh { shift->{dbh} }

sub run {
	my $self = shift;
	try {
		$self->connect;
		$self->setup;
	} catch {
		warn "Failure: $_";
	};
	while(my $data = $self->sth_ch->recv) {
		try {
			my $method = $data->{op};
			my $code = $self->can($method) or die 'unknown operation';
			$self->ret_ch->send($code->($self, $data));
		} catch {
			# warn "err: $_\n";
			$self->ret_ch->send({ status => 'fail', message => $_ });
		};
	}
}

1;

__END__

=head1 AUTHOR

Tom Molesworth <cpan@entitymodel.com>

=head1 LICENSE

Copyright Tom Molesworth 2012-2014. Licensed under the same terms as Perl itself.


lib/DBIx/Async/Worker/SQLite.pm
package DBIx::Async::Worker::SQLite;
{
  $DBIx::Async::Worker::SQLite::VERSION = '0.002';
}
use strict;
use warnings;
use parent qw(DBIx::Async::Worker);

use constant DEBUG => 0;

# Just a no-op, if anything this should be configurable behaviour.
sub setup {
	my $self = shift;
	my $dbh = $self->dbh;
	if(0) {
		# This doesn't really serve any valid purpose with
		# sqlite running in a separate process, but it
		# can make writes more predictable.
		warn "Enable journal mode...\n" if DEBUG;
		$dbh->do(q{PRAGMA journal_mode=WAL});
		warn "Disable autocheckpoint...\n" if DEBUG;
		$dbh->do(q{PRAGMA wal_autocheckpoint=0});
		warn "Switch to sync=NORMAL...\n" if DEBUG;
		$dbh->do(q{PRAGMA synchronous=NORMAL});

		$dbh->sqlite_commit_hook(sub {
			warn "Manual checkpoint...\n" if DEBUG;
			my $sth = $dbh->prepare(q{PRAGMA wal_checkpoint(FULL)});
			$sth->execute;
			while(my $row = $sth->fetchrow_arrayref) {
				warn "Checkpoint result: @$row\n" if DEBUG;
			}
			warn "Done\n" if DEBUG;
			0
		});
	}
}

1;

__END__

=head1 AUTHOR

Tom Molesworth <cpan@entitymodel.com>

=head1 LICENSE

Copyright Tom Molesworth 2012-2014. Licensed under the same terms as Perl itself.



Makefile.PL

# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v5.009.
use strict;
use warnings;

use 5.010001;

use ExtUtils::MakeMaker 6.48;



my %WriteMakefileArgs = (
  "ABSTRACT" => "database support for IO::Async via DBI",
  "AUTHOR" => "Tom Molesworth <cpan\@entitymodel.com>",
  "BUILD_REQUIRES" => {},
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => "6.48"
  },
  "DISTNAME" => "DBIx-Async",
  "EXE_FILES" => [],
  "LICENSE" => "perl",
  "NAME" => "DBIx::Async",
  "PREREQ_PM" => {
    "DBI" => 0,
    "Future" => "0.21",
    "IO::Async" => "0.60",
    "Try::Tiny" => 0,
    "curry" => 0,
    "parent" => 0
  },
  "TEST_REQUIRES" => {
    "File::Spec" => 0,
    "IO::Handle" => 0,
    "IPC::Open3" => 0,
    "Test::CheckDeps" => "0.010",
    "Test::Fatal" => "0.010",
    "Test::More" => "0.98"
  },
  "VERSION" => "0.002",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


my %FallbackPrereqs = (
  "DBI" => 0,
  "File::Spec" => 0,
  "Future" => "0.21",
  "IO::Async" => "0.60",
  "IO::Handle" => 0,
  "IPC::Open3" => 0,
  "Test::CheckDeps" => "0.010",
  "Test::Fatal" => "0.010",
  "Test::More" => "0.98",
  "Try::Tiny" => 0,
  "curry" => 0,
  "parent" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);



